src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -364,34 +373,262 @@ public void WriteValue(sbyte value)
+                   sourceChunk,
+                   destination,
+                   out int charsConsumed,
+                   out int charsWritten,
+                   isFinalBlock
+               );
+
+            Debug.Assert(status == OperationStatus.Done || status == OperationStatus.NeedMoreData);
+
+            preChunkNotProcessed = sourceChunk.Length - charsConsumed;
+            WriteStringChunk(destination.Slice(0, charsWritten), isFinalBlock);
+
+            if (valueArray != null)
+            {
+                ArrayPool<char>.Shared.Return(valueArray);
+            }
+        }
+
+        /// <summary>
+        /// Writes a chunk of the string into the buffer, converting it to UTF-8 encoding.
+        /// </summary>
+        /// <param name="chunk">The chunk of the string to be written into the buffer.</param>
+        /// <param name="isFinalBlock">Indicates whether the chunk is the final block of the string.</param>
+        /// <remarks>
+        /// This method converts the provided chunk of the string from UTF-16 encoding to UTF-8 encoding
+        /// and writes it into the buffer. It ensures that the buffer has sufficient space for the encoded
+        /// string and notifies the buffer writer of the write operation. This method assumes that the UTF-8
+        /// encoding does not exceed the maximum byte count specified by the UTF-8 encoding for the given
+        /// chunk of the string.
+        /// </remarks>
+        private void WriteStringChunk(ReadOnlySpan<char> chunk, bool isFinalBlock)
+        {
+            int maxUtf8Length = Encoding.UTF8.GetMaxByteCount(chunk.Length);
+
+            Span<byte> output = bufferWriter.GetSpan(maxUtf8Length);
+            OperationStatus status = Utf8.FromUtf16(chunk, output, out int charsRead, out int charsWritten, isFinalBlock);
+            Debug.Assert(status == OperationStatus.Done || status == OperationStatus.NeedMoreData);
+            Debug.Assert(charsRead == chunk.Length);
+
+            // notify the bufferWriter of the write
+            bufferWriter.Advance(charsWritten);
+        }
+
+        /// <summary>
+        /// Determines whether a separator should be written before the next value.
+        /// </summary>
+        /// <remarks>
+        /// This method checks the current writing context to decide whether a separator should be written
+        /// before the next value in the JSON output. It sets the flag indicating whether a separator should
+        /// be written based on conditions such as whether the writer is at the start of an array, writing
+        /// consecutive raw values at the start of an array, or if the writer is not currently in an array.
+        /// </remarks>
+        private void CheckIfSeparatorNeeded()
+        {
+            if (this.isWritingAtStartOfArray || this.isWritingConsecutiveRawValuesAtStartOfArray || !this.IsInArray())
+            {
+                this.shouldWriteSeparator = true;
+            }

```

Comment:
no else clause to set the 'shouldWriteSeparator' = false; ?

Reply:
can we find a better way to do it? 
It seems you have to change other methods to 'check shouldWriteSeparator' first, then do other things?

Reply:
Yes, I also wish we can find a better away. I introduced these separator flags in a previous PR. The challenge here is that we are bypassing `Utf8JsonWriter`. `Utf8JsonWriter` does not know that we have written these values, and so it will not write the separate when we call it next time. The result is that we'll end up with invalid JSON. To avoid that issue, we keep track of whether to write the separator ourselves when we write to the buffer manually. This means we have to coordinate with other methods of this class to write the separator before they write the value. It's not ideal, but it's the workaround we have at the moment. We use extensive tests to ensure we do this correctly.

The best way around this is for `Utf8JsonWriter` to provide these features out of the box so that we don't have to implement them ourselves. .NET 6+ added a native implementation of `WriteRawValue`, so we can use that when we upgrade. However they've not provided an ETA for when they'll address the issue with large values and chunking, but they have an open issue: https://github.com/dotnet/runtime/issues/67337